xdsl/rewriting/composable_rewriting/immutable_ir/immutable_ir.py

```diff
@@ -0,0 +1,550 @@
+                mutable_successors.append(block_mapping[successor])
+            else:
+                raise Exception("Block used before definition")
+
+        mutable_regions: List[Region] = []
+        for region in self.regions:
+            mutable_regions.append(
+                region.get_mutable_copy(value_mapping=value_mapping,
+                                        block_mapping=block_mapping))
+
+        new_op: Operation = self.op_type.create(
+            operands=mutable_operands,
+            result_types=[result.typ for result in self.results],
+            attributes=self.attributes.copy(),
+            successors=mutable_successors,
+            regions=mutable_regions)
+
+        for idx, result in enumerate(self.results):
+            m_result = new_op.results[idx]
+            value_mapping[result] = m_result
+
+        return new_op
+
+    @classmethod
+    def from_mutable(
+            cls,
+            op: Operation,
+            value_map: Optional[dict[SSAValue, ISSAValue]] = None,
+            block_map: Optional[dict[Block, IBlock]] = None,
+            existing_operands: Optional[Sequence[ISSAValue]] = None) -> IOp:
+        """
+        Returns an immutable operation that is a copy of the given mutable operation.
+        The value_map and block_map are used to map already known correspondings
+        of mutable values to immutable values and mutable blocks to immutable blocks.
+        """
+        assert isinstance(op, Operation)
+        op_type = op.__class__
+
+        if value_map is None:
+            value_map = {}
+        if block_map is None:
+            block_map = {}
+
+        operands: List[ISSAValue] = []
+        if existing_operands is None:
+            for operand in op.operands:
+                match operand:
+                    case OpResult():
+                        if operand in value_map:
+                            operands.append(value_map[operand])
+                        else:
+                            raise Exception("Operand used before definition")
+                    case BlockArgument():
+                        if operand not in value_map:
+                            raise Exception(
+                                "Block argument expected in mapping for op: " +
+                                op.name)
+                        operands.append(value_map[operand])
+                    case _:
+                        raise Exception(
+                            "Operand is expected to be either OpResult or BlockArgument"
+                        )
+        else:
+            operands.extend(existing_operands)
+
+        attributes: dict[str, Attribute] = op.attributes.copy()
+
+        successors: List[IBlock] = []
+        for successor in op.successors:
+            if successor in block_map:
+                successors.append(block_map[successor])
+            else:
+                # TODO: I think this is not right, build tests with successors
```

Comment:
From my understanding, this should never run, since (dummy) blocks should have been created at that point.

Reply:
I think the only possible case is when `from_mutable` is called on an op in a region without calling it on the parent op. I replaced the else branch here with an Exception.

Reply:
I don't see the exception. Did you forget to push at some point?

Reply:
Indeed, I did now